PERFORMANCE.md

```diff
@@ -0,0 +1,59 @@
+# Performance
+
+It is important to be conscious of performance in hot code paths. In such code paths there should be 3 goals:
+
+- Be as fast as possible
+- Be memory efficient
+- Be as stable as possible
+
+This guide is meant to document known performance issues so that the contributors are aware of them.
+
+## Performance Sensitive Areas
+
+- Change detection: `observer`, `computed`
+- View creation: `View`, `Renderer`
+- Routing: `Router`
+- Dependency Injection: `Container`, `Registry`
+
+__Note :__ If you are not in a performance sensitive area feel free to use these constructs. That being said, please do your due diligence to ensure your contribution is indeed not in a hot path.
+
+## Array Methods
+
+Most of the time using array methods e.g. `forEach`, `map`, `filter`, is fine, but in hot paths these methods can become very costly. One would think that these methods would almost be like macros that would compile into more efficient code, but that is not the case. In hot paths these methods can be up to 10X slower than just writing a `for`-loop.
+
+The reason behind the slowness is largely do to the fact that these methods take a closure which allocates memory. In turn this creates GC pressure and potentially creates a polymorphic call site.
+
+## Closures
+
+Try to keep closures to a minimum in hot code paths.
+
+- They create GC pressure.
+- They may result in polymorphic call site (see below).
+
+The reason behind why closures are slower is that contrary to popular belief, closures actually cause the VM to create a stateful `context` object at the call site. This is very similar to `new`-ing a constructor and creating an instance at the call site. Because of this closures are grouped into the same Polymorphic and Monomorphic call site heuristics.
+
+## Polymorphic vs Monomorphic Call Sites
```

Comment:
likely also want to through in megamorphic vs polymorphic.

TL:DR JITs can handle polymorphic (only a few different `fn`'s) scenarios more reasonably, but if `fn` are always changing that can cause grief.

Also, if there are 4 different shapes, and we encounter 3 always, but only sometimes encounter the fourth, that will cause a deopt, forcing potentially large swaths of compiled code to be evicted, the resulting re-compilation will be costly, and it may not correctly recover.

Priming the IC's is a way to prevent ^.

It is also worth noting, polymorphism/megamorpish can sometimes be unavoidable, but with some thought the impact can be reduced and mitigated.